c/src/lib/libbsp/arm/shared/stm32fxxxx/uart/hal-uart.c

```diff
@@ -0,0 +1,609 @@
+  if ( pUart->base_driver_info.uart_mode != STM32F_UART_MODE_POLLING ) {
+
+    if ( ret == RTEMS_SUCCESSFUL ) {
+      ret = rtems_interrupt_handler_remove(
+        pUart->base_driver_info.interrupt_number,
+        stm32f_uart_isr,
+        &pUart->base_driver_info
+        );
+    }
+  }
+
+  return (ret == RTEMS_SUCCESSFUL);
+}
+
+static int uart_init(
+  stm32_uart_driver_entry * pUart,
+  uint32_t baud
+)
+{
+  rtems_status_code ret;
+
+  //Configure the UART peripheral
+  pUart->base_driver_info.handle->Instance =
+    stmf32_uart_get_registers(pUart->base_driver_info.uart);
+  pUart->base_driver_info.handle->Init.BaudRate = pUart->base_driver_info.baud;
+  pUart->base_driver_info.handle->Init.WordLength = UART_WORDLENGTH_8B;
+  pUart->base_driver_info.handle->Init.StopBits = UART_STOPBITS_1;
+  pUart->base_driver_info.handle->Init.Parity = UART_PARITY_NONE;
+  pUart->base_driver_info.handle->Init.HwFlowCtl = UART_HWCONTROL_NONE;
+  pUart->base_driver_info.handle->Init.Mode = UART_MODE_TX_RX;
+  pUart->base_driver_info.handle->Init.OverSampling = UART_OVERSAMPLING_16;
+
+  // Initialize UART pins, clocks, and DMA controllers
+  if ( HAL_UART_Init(pUart->base_driver_info.handle) == HAL_OK ) {
+    ret = RTEMS_SUCCESSFUL;
+  }
+  else {
+    ret = RTEMS_UNSATISFIED;
+  }
+
+  return ret;
+}
+
+static int uart_de_init(
+  stm32_uart_driver_entry * pUart
+)
+{
+  rtems_status_code ret;
+
+  if ( HAL_UART_DeInit(pUart->base_driver_info.handle) != HAL_OK ) {
+    ret = RTEMS_UNSATISFIED;
+  }
+  else {
+    ret = RTEMS_SUCCESSFUL;
+  }
+
+  return ret;
+}
+
+static void uart_destroy(
+  stm32_uart_driver_entry * pUart
+)
+{
+  (void) uart_de_init(pUart);
+}
+
+static const IMFS_node_control uart_node_control =
+  IMFS_GENERIC_INITIALIZER(
+    &uart_handler,
+    IMFS_node_initialize_generic,
+    uart_node_destroy
+    );
+
+static int uart_register_device_driver(
+  stm32_uart_device * pUartDevice
+)
+{
+  int rv;
+  rtems_status_code sc;
+
+  rv = IMFS_make_generic_node(
+    pUartDevice->pUart->base_driver_info.device_name,
+    S_IFCHR | S_IRWXU | S_IRWXG | S_IRWXO,
+    &uart_node_control,
+    pUartDevice
+    );
+
+  if ( rv != 0 ) {
+
+    (*pUartDevice->destroy)(pUartDevice->pUart);
+
+  } else {
+
+    sc = rtems_task_start(
+      pUartDevice->pUart->tx_task_id,
+      pUartDevice->pUart->tx_task,
+      (rtems_task_argument) pUartDevice->pUart
+      );
+  }
+  return sc;
+}
+
+static int uart_create_rtems_objects(
+  stm32_uart_device *pUartDevice
+)
+{
+  rtems_status_code sc;
+
+  char uart_num = '1' + (char) pUartDevice->pUart->base_driver_info.uart;
+
+  sc = rtems_message_queue_create(
+    rtems_build_name('U', 'T', 'X', uart_num),
+    UART_TX_QUEUE_LEN,
+    MAX_UART_TX_MSG_SIZE,
+    RTEMS_FIFO | RTEMS_LOCAL,
+    &pUartDevice->pUart->tx_msg_queue
+    );
+
+  if ( sc != RTEMS_SUCCESSFUL ) {
+    uart_destroy(pUartDevice->pUart);
+    rtems_set_errno_and_return_minus_one(ENOMEM);
+  }
+
+  sc = rtems_task_create(
```

Comment:
I think we can make both the CAN and UART without using tasks. 
I haven't fully thought this out, but we could probably do something similar to what the PIC32 is doing, with a message queue and some clever flags.. This is just a thought though, not necessarily true though..